janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java

```diff
@@ -114,8 +115,33 @@ public void setup(final Map<String,Object> config) {
         }
 
         final String defaultUser = config.get(CONFIG_DEFAULT_USER).toString();
-        if (credentialStore.findUser(defaultUser) == null) {
-            credentialStore.createUser(defaultUser, config.get(CONFIG_DEFAULT_PASSWORD).toString());
+        if (!userExists(defaultUser)) {
```

Comment:
We used to use a check like `credentialStore.findUser(defaultUser) == null` here which throws an exception if multiple users exist. Now we don't throw this exception if multiple users exist because we use `userExists(defaultUser)`. Is it OK?

Reply:
Good find! I'll modify `userExists()` to also throw an exception in that case to not change the behaviour here.

Reply:
After some more investigation I noticed that it's not possible to get duplicate users because we're using a uniqueness constraint on the `username` property. I explained this a bit further in [your issue about the race condition](https://github.com/JanusGraph/janusgraph/issues/1434#issuecomment-465906853).

So, I actually went the opposite way and removed the duplicate check in `findUser()` instead.

I pushed that with a separate commit so you can easier review it and I will squash them when you're done with the review.

Reply:
@FlorianHockmann I answered about commented in the issue that I am not sure how it will work with eventual consistency because that code doesn't use locks. I.e. I think we should use something like:
```
mgmt.setConsistency(name, ConsistencyModifier.LOCK)
mgmt.setConsistency(index, ConsistencyModifier.LOCK)
```
but I am not sure. We need to test it better. 

As this isn't related to the current PR, I think right now it will be better to not change a flow (I.e. throw exception in `findUser` and `userExists` methods) and then after some investigation open a new PR (if needed) to remove duplicate checks.

What do you think?

Reply:
I'm not sure what you mean with _that code doesn't use locks_. Sure, the code itself doesn't set any locks, but JanusGraph will acquire a lock in the backend to modify anything that is guarded by a uniqueness constraint. This is basically distributed locking backed by the storage backend.
So, you won't be able to add two vertices with the same `username` property, even if you try it in parallel from different threads / JanusGraph instances.

Reply:
In the issue you said that we use `uniqueness constraint`. And you pointed to this part of code:
https://github.com/JanusGraph/janusgraph/blob/3812806a98598965950c366f1613210091330b64/janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java#L99

I checked that code and see that we don't use locks there. I.e. we don't use `mgmt.setConsistency(name, ConsistencyModifier.LOCK)`.

Yes, we use `uniqueness constraint` but I pointed to [this documentation](https://docs.janusgraph.org/latest/eventual-consistency.html) which explains that we must obtain locks to ensure consistency. As I said, I don't know how `uniqueness constraint` works in eventual consistent backend without locks. If you know that with `unique` constraint we won't be able to create duplicate user without `ConsistencyModifier.LOCK` in eventual consistent backend then I am OK with the current change. If you are not sure then I think we should not remove duplicate checks in your PR.

Reply:
Ah, now I see what you mean. My understanding was always that making an index `unique` is already enough to enforce uniqueness of its indexed properties. But it looks like I was wrong about that and you actually have to enable locking explicitly for eventually consistent storage backends like it's shown in the example you linked to in the docs.

This is also explained[ here](https://docs.janusgraph.org/0.3.1/indexes.html#_composite_index) in the docs:

> To enforce uniqueness against an eventually consistent storage backend, the consistency of the index must be explicitly set to enabling locking.

So, I agree with you that I should not remove the existence checks with this PR. I'll revert my last commit here and instead add another check in `userExists()` as you suggested initially.
I suggest that we enable locking explicitly for eventually consistent storage backends in the issue you created.

Reply:
Done